cassiopeia/ProcessingPipeline/process/pipeline.py

```diff
@@ -98,33 +100,150 @@ def resolve_UMI_sequence(
+    mismatches and indels in the sequence read itself. Writes out a dataframe of the collapsed UMIs table.
+
+    Args:
+          out_dir: The output directory where the sorted bam directory, the collapsed bam directory, and the final collapsed table are written to.
+          bam_file_name: File path of the bam_file. Just the bam file name can be specified if the bam already exists in the output directory.
+          max_hq_mismatches: A threshold specifying the maximum number of high quality mismatches between the seqeunces of 2 aligned segments to be collapsed.
+          max_indels: A threshold specifying the maximum number of differing indels allowed between the sequences of 2 aligned segments to be collapsed.
+          n_threads: Number of threads used. Currently only supports single threaded use.
+          show_progress: Allow progress bar to be shown.
+          force_sort: Specify whether to sort the initial bam directory, regardless of if the sorted file already exists.
+
+    Returns:
+          None; output table is written to file, which can then be read in as a dataframe.
+    """
+
+    logging.info("Collapsing UMI sequences...")
+
+    t0 = time.time()
+
+    # pathing written such that the bam file that is being converted does not have to exist currently in the output directory
+    if out_dir[-1] == "/":
+        out_dir = out_dir[:-1]
+    sorted_file_name = Path(
+        out_dir
+        + "/"
+        + ".".join(bam_file_name.split("/")[-1].split(".")[:-1])
+        + "_sorted.bam"
+    )
+
+    sort_key = lambda al: (al.get_tag(CELL_BC_TAG), al.get_tag(UMI_TAG))
+    filter_func = lambda al: al.has_tag(CELL_BC_TAG)
+
+    if force_sort or not sorted_file_name.exists():
+        max_read_length, total_reads_out = UMI_utils.sort_cellranger_bam(
+            bam_file_name,
+            str(sorted_file_name),
+            sort_key,
+            filter_func,
+            show_progress=show_progress,
+        )
+        logging.info("Sorted bam directory saved to " + str(sorted_file_name))
+        logging.info("Max read length of " + str(max_read_length))
+        logging.info("Total reads: " + str(total_reads_out))
+
+    collapsed_file_name = sorted_file_name.with_suffix(".collapsed.bam")
+    if not collapsed_file_name.exists():
+        UMI_utils.form_collapsed_clusters(
+            str(sorted_file_name),
+            max_hq_mismatches,
+            max_indels,
+            show_progress=show_progress,
+        )
+
+    logging.info(f"Finished collapsing UMI sequences in {time.time() - t0} s.")
+    collapsed_df_file_name = sorted_file_name.with_suffix(".collapsed.txt")
+    convertBam2DF(str(collapsed_file_name), str(collapsed_df_file_name))
+    logging.info("Collapsed bam directory saved to " + str(collapsed_file_name))
+    logging.info("Converted dataframe saved to " + str(collapsed_df_file_name))
+
+
+def convertBam2DF(data_fp: str, out_fp: str):
```

Comment:
Please write some tests to check this function. We can put them in the same `test` directory that I created.